### PR TITLE
feat: job upsert

### PR DESCRIPTION
**Pull request by arinda-arif**
Adding a job upsert capability.
- Will accept multiple jobs
- Will not return error if the job is deleted and wants to get recreated in the same namespace
- Deploy command will use this upsert api

Dependency: 
- [x] Merge proton [PR](https://github.com/goto/proton/pull/107/commits/89a92faae610999b24cc9394b57b141dd3df6cf4)

### CODE REVIEW

File: core/job/service/job_service.go [hunk 1]
```diff
@@ -262,6 +262,89 @@ func (j *JobService) Update(ctx context.Context, jobTenant tenant.Tenant, specs
...
+		}
+		jobsToUpdate = append(jobsToUpdate, jobToUpsert)
+		existingJobs[jobToUpsert.Spec().Name()] = existingJob
+	}
+
+	var upsertedJobs, addedJobs, updatedJobs []*job.Job
+	if len(jobsToAdd) > 0 {
+		addedJobs, err = j.jobRepo.Add(ctx, jobsToAdd)
+		upsertedJobs = append(upsertedJobs, addedJobs...)
+		logWriter.Write(writer.LogLevelInfo, fmt.Sprintf("[%s] added %d jobs", tenantWithDetails.Namespace().Name().String(), len(addedJobs)))
```
Comment (deryrahman) [inline core/job/service/job_service.go:298]: can we omit the log if the added/updated jobs is 0?

File: core/job/service/job_service.go [hunk 2]
```diff
@@ -983,9 +1066,10 @@ func (j *JobService) generateJobs(ctx context.Context, tenantWithDetails *tenant
 			return func() (interface{}, error) {
 				generatedJob, err := j.generateJob(ctx, tenantWithDetails, currentSpec)
 				if err != nil {
-					j.logger.Error("error generating job [%s]: %s", currentSpec.Name(), err)
+					errorMsg := fmt.Sprintf("error generating job [%s]: %s", spec.Name().String(), err.Error())
```
Comment (deryrahman) [inline core/job/service/job_service.go:1069]: is it expected to change `currentSpec` to `spec`? in this case `spec` is outside the function, doesn't it lead to unpredicted behavior when running concurrently?

File: core/job/service/job_service.go (hunk 1 above)
Comment (arinda-arif) [reply core/job/service/job_service.go:298]: sure

File: core/job/service/job_service.go (hunk 2 above)
Comment (arinda-arif) [reply core/job/service/job_service.go:1069]: yes, it should be `currentSpec`. thank you for noticing this.

Comment (arinda-arif) [reply core/job/service/job_service.go]: updated

File: core/job/service/job_service.go (hunk 1 above)
Comment (arinda-arif) [reply core/job/service/job_service.go]: updated

File: core/job/handler/v1beta1/job.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ func NewJobHandler(jobService JobService, changeLogService ChangeLogService, log
 type JobService interface {
 	Add(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
 	Update(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
+	Upsert(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
```
Comment (sbchaos) [inline core/job/handler/v1beta1/job.go:49]: Can we break this interface further, it already has too many methods.

For example one such breakdown can be.
If you do not want to make changes at the moment just break the interface and use components in JobService interface

eg
```go
type JobGetter interface {
        Get(ctx context.Context, jobTenant tenant.Tenant, jobName job.Name) (jobSpec *job.Job, err error)
	GetTaskInfo(ctx context.Context, task job.Task) (*plugin.Info, error)
	GetByFilter(ctx context.Context, filters ...filter.FilterOpt) (jobSpecs []*job.Job, err error)
	GetJobBasicInfo(ctx context.Context, jobTenant tenant.Tenant, jobName job.Name, spec *job.Spec) (*job.Job, writer.BufferedLogger)
	GetUpstreamsToInspect(ctx context.Context, subjectJob *job.Job, localJob bool) ([]*job.Upstream, error)
	GetDownstream(ctx context.Context, job *job.Job, localJob bool) ([]*job.Downstream, error)
}

type JobModificationService interface {
        Add(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
 	Update(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
 	Upsert(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
         ChangeNamespace(ctx context.Context, jobSourceTenant, jobNewTenant tenant.Tenant, jobName job.Name) error
	Delete(ctx context.Context, jobTenant tenant.Tenant, jobName job.Name, cleanFlag, forceFlag bool) (affectedDownstream []job.FullName, err error)
}

type JobStatusService interface {
        ReplaceAll(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec, jobNamesWithInvalidSpec []job.Name, logWriter writer.LogWriter) error
	Refresh(ctx context.Context, projectName tenant.ProjectName, namespaceNames, jobNames []string, logWriter writer.LogWriter) error
	Validate(context.Context, dto.ValidateRequest) (map[job.Name][]dto.ValidateResult, error)
        SyncState(ctx context.Context, jobTenant tenant.Tenant, disabledJobNames, enabledJobNames []job.Name) error
 	UpdateState(ctx context.Context, jobTenant tenant.Tenant, jobNames []job.Name, jobState job.State, remark string) error
}

type JobService interface {
      JobGetterService
      JobModificationService
      JobStatusService
}

```

File: core/job/handler/v1beta1/job.go [hunk 4]
```diff
@@ -46,6 +46,7 @@ func NewJobHandler(jobService JobService, changeLogService ChangeLogService, log
 type JobService interface {
 	Add(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
 	Update(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
+	Upsert(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]dto.UpsertResult, error)
```
Comment (sbchaos) [inline core/job/handler/v1beta1/job.go:49]: Can we break this interface further, it already has too many methods.

For example one such breakdown can be.
If you do not want to make changes at the moment just break the interface and use components in JobService interface

eg

```go
type JobQueryService interface {
        Get(ctx context.Context, jobTenant tenant.Tenant, jobName job.Name) (jobSpec *job.Job, err error)
	GetTaskInfo(ctx context.Context, task job.Task) (*plugin.Info, error)
	GetByFilter(ctx context.Context, filters ...filter.FilterOpt) (jobSpecs []*job.Job, err error)
	GetJobBasicInfo(ctx context.Context, jobTenant tenant.Tenant, jobName job.Name, spec *job.Spec) (*job.Job, writer.BufferedLogger)
	GetUpstreamsToInspect(ctx context.Context, subjectJob *job.Job, localJob bool) ([]*job.Upstream, error)
	GetDownstream(ctx context.Context, job *job.Job, localJob bool) ([]*job.Downstream, error)
}

type JobModificationService interface {
        Add(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
 	Update(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
 	Upsert(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec) ([]job.Name, error)
        ChangeNamespace(ctx context.Context, jobSourceTenant, jobNewTenant tenant.Tenant, jobName job.Name) error
	Delete(ctx context.Context, jobTenant tenant.Tenant, jobName job.Name, cleanFlag, forceFlag bool) (affectedDownstream []job.FullName, err error)
}

type JobStatusService interface {
        ReplaceAll(ctx context.Context, jobTenant tenant.Tenant, jobs []*job.Spec, jobNamesWithInvalidSpec []job.Name, logWriter writer.LogWriter) error
	Refresh(ctx context.Context, projectName tenant.ProjectName, namespaceNames, jobNames []string, logWriter writer.LogWriter) error
	Validate(context.Context, dto.ValidateRequest) (map[job.Name][]dto.ValidateResult, error)
        SyncState(ctx context.Context, jobTenant tenant.Tenant, disabledJobNames, enabledJobNames []job.Name) error
 	UpdateState(ctx context.Context, jobTenant tenant.Tenant, jobNames []job.Name, jobState job.State, remark string) error
}

type JobService interface {
      JobQueryService
      JobModificationService
      JobStatusService
}
```

Comment (arinda-arif) [reply core/job/handler/v1beta1/job.go:49]: sure

File: core/job/handler/v1beta1/job.go (hunk 3 above)
Comment (arinda-arif) [reply core/job/handler/v1beta1/job.go]: updated

File: core/job/handler/v1beta1/job.go (hunk 4 above)
Comment (arinda-arif) [reply core/job/handler/v1beta1/job.go:49]: updated

**Review by deryrahman — APPROVED** [review]
LGTM

**Review by sbchaos — APPROVED** [review]

**Pull request merged**